File: Backend/dependency/inquiry/src/main/java/com/mhp/coding/challenges/dependency/inquiry/InquiryService.java
package com.mhp.coding.challenges.dependency.inquiry;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.beans.factory.annotation.Autowired;

import org.springframework.context.ApplicationEventPublisher;

/*
 * This tutorial to Spring events was used to solve the challenge: 
 * https://www.baeldung.com/spring-events
 */
@Component
public class InquiryService {

    private static final Logger LOG = LoggerFactory.getLogger(InquiryService.class);
    
    // Publisher for the custom ApplicationEvent
    // use @Autowired for automatic dependency injection by Spring
    @Autowired
    private ApplicationEventPublisher eventPublisher;

    public void create(final Inquiry inquiry) {
        LOG.info("User sent inquiry: {}", inquiry);
        
        // create an event that contains the inquiry for the listeners,
        // i.e. the PushNotificationHandler and the EmailHandler
        InquiryServiceEvent event = new InquiryServiceEvent(this, inquiry);
        // publish the event
        eventPublisher.publishEvent(event);
    }

}

File: Backend/dependency-kotlin/inquiry/src/main/kotlin/com/mhp/coding/challenges/dependency/inquiry/InquiryService.kt
package com.mhp.coding.challenges.dependency.inquiry

import mu.KotlinLogging
import org.springframework.stereotype.Component
import org.springframework.beans.factory.annotation.Autowired;

import org.springframework.context.ApplicationEvent;
import org.springframework.context.ApplicationEventPublisher;

private val logger = KotlinLogging.logger {}

@Component
class InquiryService(
    @Autowired private val eventPublisher : ApplicationEventPublisher
) {
    
    fun create(inquiry: Inquiry) {
        logger.info {
            "User sent inquiry: $inquiry"
        }
        // create an event that contains the inquiry for the listeners,
        // i.e. the PushNotificationHandler and the EmailHandler
        val event = InquiryServiceEvent(
            inquiry = inquiry,
            eventSource = this
        )
        // publish the event
        eventPublisher.publishEvent(event)
    }
}

/*
 * This class is used to create custom Spring ApplicationEvents.
 * The inquiry an event is connected to is contained in the inquiry property.
 */
class InquiryServiceEvent : ApplicationEvent {
    
    val inquiry : Inquiry
    
    constructor(inquiry : Inquiry, eventSource : Any) : super(eventSource) { 
        this.inquiry = inquiry 
    }
}

data class Inquiry(
    var username: String,
    var recipient: String,
    var text: String,
)
